src/hdmf/common/resources.py

```diff
@@ -459,99 +516,109 @@ def add_ref(self, **kwargs):
             key = self._add_key(key)
             self._add_object_key(object_field, key)
 
-        if kwargs['resources_idx'] is not None and kwargs['resource_name'] is None and kwargs['resource_uri'] is None:
-            resource_table_idx = kwargs['resources_idx']
-        elif (
-            kwargs['resources_idx'] is not None
-            and (kwargs['resource_name'] is not None
-                 or kwargs['resource_uri'] is not None)):
-            msg = "Can't have resource_idx with resource_name or resource_uri."
-            raise ValueError(msg)
-        elif len(self.resources.which(resource=kwargs['resource_name'])) == 0:
-            resource_name = kwargs['resource_name']
-            resource_uri = kwargs['resource_uri']
-            resource_table_idx = self._add_resource(resource_name, resource_uri)
-        else:
-            idx = self.resources.which(resource=kwargs['resource_name'])
-            resource_table_idx = self.resources.row[idx[0]]
-
-        if (resource_table_idx is not None and entity_id is not None and entity_uri is not None):
-            add_entity = True
-        elif not (resource_table_idx is None and entity_id is None and resource_uri is None):
-            msg = ("Specify resource, entity_id, and entity_uri arguments."
-                   "All three are required to create a reference")
-            raise ValueError(msg)
+        entity = self._add_entity(key, entity_id, entity_uri)
+
+        return key, entity
+
+    @docval({'name': 'object_type', 'type': str,
+             'doc': 'The type of the object. This is also the parent in relative_path.'},
+            {'name': 'relative_path', 'type': str,
+             'doc': ('The relative_path of the attribute of the object that uses ',
+                     'an external resource reference key. Use an empty string if not applicable.'),
+             'default': ''},
+            {'name': 'field', 'type': str, 'default': '',
+             'doc': ('The field of the compound data type using an external resource.')},
+            {'name': 'all_instances', 'type': bool, 'default': False,
+             'doc': ('The bool to return all a dataframe with all instances of the object_type.',
+                     'If True, relative_path and field inputs will be ignored.')})
+    def get_object_type(self, **kwargs):
```

Comment:
Could you please remind me, what is the use case for this method? I have a vague memory that we all agreed on this but I forget why. Because `object_type = container.__class__.__name__`, I worry that this will not do what we want in some cases. For example, `get_object_type("Data")` will return only objects that have neurodata type "Data" and not objects that are of a subclass of TimeSeries. It also sets up a situation where the data have to be kept consistent (i.e., what do we do if the object type does not match the object?). That constraint might be worth it though, depending on the value gained for the user.